grpclib/protocol.py

```diff
@@ -223,13 +230,88 @@ def flush(self) -> None:
         if data:
             self._transport.write(data)
 
+    def initialize(self) -> None:
+        if self._config._keepalive_time is not None:
+            self._ping_handle = asyncio.get_event_loop().call_later(
+                self._config._keepalive_time,
+                self._ping
+            )
+
     def close(self) -> None:
         if hasattr(self, '_transport'):
             self._transport.close()
             # remove cyclic references to improve memory usage
             del self._transport
             if hasattr(self._connection, '_frame_dispatch_table'):
                 del self._connection._frame_dispatch_table
+        if self._ping_handle is not None:
+            self._ping_handle.cancel()
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+
+    def _ping(self) -> None:
+        data = struct.pack('!Q', int(time.monotonic() * 10 ** 6))
+        is_need_send = True
+
+        if not self._config._keepalive_permit_without_calls:
+            count_of_connection = any(
+                s.open for s in self._connection.streams.values()
+            )
+            if count_of_connection == 0:
+                is_need_send = False
+
+        _keepalive_time = cast(float, self._config._keepalive_time)
+        current_time = time.monotonic()
+        if current_time - self._last_received() < _keepalive_time:
+            is_need_send = False
+
+        if self.last_ping_sent is not None and \
+                current_time - self.last_ping_sent < \
+                self._config._http2_min_sent_ping_interval_without_data:
+            is_need_send = False
+
+        if self._config._http2_max_pings_without_data != 0 and \
+                self.ping_count_in_sequence >= \
+                self._config._http2_max_pings_without_data:
+            is_need_send = False
+
+        if is_need_send:
+            logging.error(f'send ping {self._config._keepalive_timeout}')
```

Comment:
It is better to use custom logger, e.g.:
```python
log = logging.getLogger(__name__)  # on top of the module
...
log.error(...)  # inside a function
```
And you probably want to use `log.debug` instead of `log.error`?

And another thing to consider is that you have to use %-style formatting for logging, e.g.:
```python
log.info('some message from %s with %r', username, message)
```

Reply:
This one remained in error. But now it seems to me that it is worth leaving. I will do it through "log".

Reply:
done